Source/Mothership/MPIHandlers.cpp

```diff
@@ -0,0 +1,497 @@
+/* This source file defines the Mothership methods for handling incoming MPI
+ * messages. The functionality of these handlers is described in the Mothership
+ * documentation.
+ *
+ * Each handler is called by one of two sources (disambiguated by the dyadic
+ * nature of their signature):
+ *
+ * - MPISpinner (and Decode): These either handle exit messages, or push
+ *    messages into either MPICncQueue or MPIAppQueue.
+ *
+ * - Consumers of the above queues, which updates application states.
+ *
+ * See Mothership.h for the taxonomy. */
+
+#include "Mothership.h"
+
+unsigned Mothership::handle_msg_exit(PMsg_p* message, unsigned commIndex)
+{
+    debug_post(592, 2, "Q::EXIT", "Exiting gracefully.");
+    threading.set_quit();
+
+    /* CommonBase's OnExit (does some MPI teardown). The returncode is read by
+     * Decode, and causes MPISpinner to end. */
+    OnExit(message, commIndex);
+    return 1;
+}
+
+unsigned Mothership::handle_msg_syst_kill(PMsg_p* message, unsigned commIndex)
+{
+    /* ThreadComms notices that we haven't set_quit, so it doesn't wait for the
+     * other threads to finish before leaving. */
+    debug_post(592, 2, "Q::SYST,Q::KILL", "Exiting as quickly as possible.");
+
+    /* CommonBase's OnExit (does some MPI teardown). The returncode is read by
+     * Decode, and causes MPISpinner to end. */
+    OnExit(message, commIndex);
+    return 1;
+}
+
+unsigned Mothership::handle_msg_app(PMsg_p* message, unsigned commIndex)
+{
+    #if ORCHESTRATOR_DEBUG
```

Comment:
In my experience it hasn't been an issue, because messages from each process don't tend to interleave. It's easy enough to do though (just set a condition on the flag in the `Makefile`), but the cli arguments will be weird.